### PR TITLE
nektar 4.3.1: new formula

### PR DESCRIPTION
**Pull request by mdave**
This PR adds a Formula for the Nektar++ spectral/hp element framework (www.nektar.info). To the best of my knowledge, I've complied with all of the guidelines. A slight grey area is that we use CMake's ExternalProject to download and compile modified third party software. I've encapsulated this using resources, so that homebrew can perform the download and place the tarballs in the appropriate location, but let me know if you think there are major issues with this.
### Have you:
- [x] Followed the guidelines in our [Contributing](https://github.com/Homebrew/homebrew-science/blob/master/.github/CONTRIBUTING.md) document?
- [x] Checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/homebrew-science/pulls) for the same update/change?
- [x] Run `brew audit --strict --online <formula>` (where `<formula>` is the name of the formula you're submitting) and corrected all errors?
- [x] Built your formula locally prior to submission with `brew install <formula>`?
### New formula: have you
- [x] Written a sensible test? (The best test is to compile and run a sample program.)

### CODE REVIEW

**Comment by mdave** [comment]
Okay, I think I've gone as far as I can with testing this for now. Would appreciate any input/comments! Some notes from my side:

* The Jenkins build fails due to #3461.
* `cmake` is listed as `:run` since my test uses our CMake configuration file. If it's not there, Travis will uninstall `cmake` before running tests, causing failure.

File: nektar.rb [hunk 1]
```diff
@@ -0,0 +1,176 @@
+class Nektar < Formula
+  # Nektar++ will, by default, download and build its ThirdParty
+  # dependencies. This download strategy has been added so that Homebrew can
+  # download resources and check hashes independently.
+  class NektarThirdPartyDownloadStrategy < CurlDownloadStrategy
```
Comment (dpo) [inline nektar.rb]: This should be defined outside of the formula class.

File: nektar.rb [hunk 2]
```diff
@@ -0,0 +1,176 @@
...
+  depends_on "fftw"   => :recommended
+  depends_on "petsc"  => :recommended
+  depends_on "scotch" => :optional
+
+  # Modified version of the METIS library
+  resource "modmetis" do
+    url "https://www.nektar.info/thirdparty/modmetis-5.1.0_2.tar.bz2",
+      :using => NektarThirdPartyDownloadStrategy
+    sha256 "c6496cb32c892b5ab1f78484040c1f8d53b48da232c7a7e085239ea5bd431392"
+  end
```
Comment (dpo) [inline nektar.rb]: This library won't work with the regular METIS?

File: nektar.rb (hunk 1 above)
Comment (mdave) [inline nektar.rb]: Have now moved this.

File: nektar.rb (hunk 2 above)
Comment (mdave) [inline nektar.rb]: Unfortunately not -- one relatively important part of the code relies on an internal METIS data structure that isn't exposed.

Comment (dpo) [inline nektar.rb]: Is there any chance to patch the `metis` formula, or better yet, to submit a patch upstream to expose the data structure in question?

**Comment by mdave** [comment]
I've updated this to v4.3.1 to fix clang 7.3 issues with METIS compilation. I would prefer to use our modified version, at least in the short term. The patch is pretty hairy, so I am not sure I can get a patch that I would be happy submitting upstream or modifying in the formula, at least in the short term. If this is a deal breaker then please let me know.

**Comment by dpo** [comment]
Does the formula install `libmetis`? If so, it should be stashed away so Homebrew doesn't put it in `/usr/local/lib` as it will conflict with the official `libmetis`.

**Comment by mdave** [comment]
All of the libraries are stored in `/usr/local/lib/nektar++-VERSION` so as to separate our third-party dependencies from the main library.

**Comment by iMichka** [comment]
Hi. There was no progress here since some time. It is partly understandable as this formula is pretty hairy, due to the metis question. If the change can not be done in metis, why not fork metis and do your own stuff, if you are not able to come to a compromise with the metis authors.

Please understand that we try to avoid exceptions and non-standard builds; because all this increases the maintenance burden. And we are an understaffed open-source project, so if we can avoid edge cases that would be great.

**Comment by mdave** [comment]
No problem. As a developer on another understaffed open-source project I can certainly relate!

I am going to close this issue for now and resubmit later. We have a branch under development to remove the requirement on the modified Metis library over to Scotch, which I believe natively supports what we need through the API. We've also removed the requirement for the (now abandoned) Loki header library, so this should remove all of the 'hairy' components for this Formula.

**Pull request closed without merging**